slmsuite/holography/analysis/files.py

```diff
@@ -0,0 +1,407 @@
+
+    Returns
+    -------
+    data : dict
+        Dictionary of the data stored in the file.
+    """
+    def recurse(group):
+        data = {}
+
+        for key in group.keys():
+            if isinstance(group[key], h5py.Group):
+                data[key] = recurse(group[key])
+            else:
+                data_ = group[key][()]
+                if decode_bytes:
+                    if isinstance(data_, bytes):
+                        data_ = bytes.decode(data_)
+                    elif np.isscalar(data_):
+                        pass
+                    elif isinstance(data_, np.ndarray) and len(data_) > 0 and isinstance(data_[0], bytes):
+                        data_ = np.vectorize(bytes.decode)(data_)
+                data[key] = data_
+
+        return data
+
+    with h5py.File(file_path, "r") as file_:
+        data = recurse(file_)
+
+    return data
+
+
+def write_h5(file_path, data, mode="w"):
+    """
+    Write data in a dictionary to an `h5 file
+    <https://docs.h5py.org/en/stable/high/file.html#opening-creating-files>`_.
+
+    Note
+    ~~~~
+    There are some limitations to what the h5 file standard can store, along with
+    limitations on what is currently implemented in this function. A few tips:
+
+    Supported types:
+
+    - Nested dictionaries which are written as h5 group hierarchy,
+    - ``None`` (though this is written as ``False``),
+    - Uniform arrays of numeric or string data.
+
+    Example unsupported types:
+
+    - Staggered arrays (an array consisting of arrays of different sizes),
+    - Non-numeric or non-string data (e.g. object),
+
+    Parameters
+    ----------
+    file_path : str
+        Full path to the file to save the data in.
+    data : dict
+        Dictionary of data to save in the file.
+    mode : str
+        The mode to open the file with.
+    """
+    def recurse(group, data):
+        for key in data.keys():
+            if isinstance(data[key], dict):
+                new_group = group.create_group(key)
+                recurse(new_group, data[key])
+            elif isinstance(data[key], str):
+                group[key] = bytes(data[key], 'utf-8')
+            elif data[key] is None:
+                group[key] = False
+            else:
+                try:
+                    array = np.array(data[key])
+                except ValueError as e:
+                    raise ValueError(
+                        "write_h5() does not support saving staggered arrays such as {}. "
+                        "Arrays must be uniform. {}".format(str(data[key]), str(e))
+                    )
+                except Exception as e:
+                    raise e
+
+                if array.dtype.char == "U":
+                    array = np.vectorize(str.encode)(array)
+
+                group[key] = array
+
+    with h5py.File(file_path, mode) as file_:
+        recurse(file_, data)
+
+
+def _read_image(path, shape, target_shape=None, angle=0, shift=(-225, -170)):
```

Comment:
If it’s a top level method it shouldn’t be preceded by an underscore?

Reply:
I'm using preceding underscore to denote private methods (and the docs ignore them). In this case, I wanted `_read_image` as a method to use in the examples, but wasn't happy with the current form/parameters to make it a public function. I don't care as much about documentation or backwards compatibility for private functions.